### PR TITLE
Add version injection at build time

### PR DESCRIPTION
**Pull request by iamsubratp**

### CODE REVIEW

File: container_images/daisy-builder/main.sh [hunk 1]
```diff
@@ -20,6 +20,20 @@ ZONE="$2"
 WORKFLOW_FILE="$3"      # Workflow to run
 GCS_OUTPUT_BUCKET="$4"  # Destination for artifacts
 
+function GENERATE_NEW_VERSION() {
```
Comment (hopkiw) [inline container_images/daisy-builder/main.sh]: nit: function names should be lower case

File: container_images/daisy-builder/main.sh [hunk 2]
```diff
@@ -20,6 +20,20 @@ ZONE="$2"
...
+  VERSION_GENERATE_CMD="/versiongenerator --token-file-path=${GITHUB_ACCESS_TOKEN} --org=${REPO_OWNER} --repo=${REPO_NAME}"
```
Comment (hopkiw) [inline container_images/daisy-builder/main.sh]: why store as a variable just to use in next line?

File: container_images/daisy-builder/main.sh [hunk 3]
```diff
@@ -20,6 +20,20 @@ ZONE="$2"
...
+  VERSION_GENERATE_OUT=$(${VERSION_GENERATE_CMD})
+
+  if [[ $? -ne 0 ]]; then
+    echo -e "could not generate version: $VERSION_GENERATE_OUT"
+  fi
+
+  pattern="Generated version: "
+  local BUILD_ID=$(echo $VERSION_GENERATE_OUT | sed "s/$pattern//")
```
Comment (hopkiw) [inline container_images/daisy-builder/main.sh]: can we remove the $pattern from generator command, or add flag to remove it? we control this stuff so we shouldn't have to be doing string parsing on what our utilities provide us

File: container_images/daisy-builder/main.sh [hunk 4]
```diff
@@ -37,8 +51,16 @@ if [[ "$JOB_TYPE" == "presubmit" ]]; then
   DAISY_VARS+=",pull_ref=pull/${PULL_NUMBER}/head"
 fi
 
+## generate buildID
+if [[ "$JOB_TYPE" == "postsubmit" ]]; then
+  BUILD_ID=$(GENERATE_NEW_VERSION)
+  DAISY_VARS+=",build_id=${BUILD_ID}"
```
Comment (hopkiw) [inline container_images/daisy-builder/main.sh]: can we change all references from 'build_id' to 'version' ? build id is something different

File: container_images/daisy-builder/main.sh (hunk 1 above)
Comment (iamsubratp) [reply container_images/daisy-builder/main.sh]: ack

File: container_images/daisy-builder/main.sh (hunk 2 above)
Comment (iamsubratp) [reply container_images/daisy-builder/main.sh]: used the variable to debug env variables. i have a echo in my local repo to verify that the env variables are correctly set.

File: container_images/daisy-builder/main.sh (hunk 3 above)
Comment (iamsubratp) [reply container_images/daisy-builder/main.sh]: done

File: container_images/daisy-builder/main.sh (hunk 4 above)
Comment (iamsubratp) [reply container_images/daisy-builder/main.sh]: done

File: container_images/daisy-builder/main.sh (hunk 2 above)
Comment (hopkiw) [reply container_images/daisy-builder/main.sh]: kk, let's just prune from the PR

Comment (iamsubratp) [reply container_images/daisy-builder/main.sh]: done

File: container_images/daisy-builder/main.sh [hunk 5]
```diff
@@ -20,6 +20,17 @@ ZONE="$2"
 WORKFLOW_FILE="$3"      # Workflow to run
 GCS_OUTPUT_BUCKET="$4"  # Destination for artifacts
 
+function generate_new_version() {
+  local VERSION=`/versiongenerator --token-file-path=${GITHUB_ACCESS_TOKEN} --org=${REPO_OWNER} --repo=${REPO_NAME}`
+
+  if [[ $? -ne 0 ]]; then
+    echo -e "could not generate version: $VERSION_GENERATE_OUT"
```
Comment (hopkiw) [inline container_images/daisy-builder/main.sh]: i don't think `$VERSION_GENERATE_OUT` is set

**Comment by hopkiw** [comment]
looks good except for one variable needs pruning or updating. thanks

File: container_images/daisy-builder/main.sh (hunk 5 above)
Comment (iamsubratp) [reply container_images/daisy-builder/main.sh]: done

File: container_images/daisy-builder/main.sh [hunk 6]
```diff
@@ -21,13 +21,13 @@ WORKFLOW_FILE="$3"      # Workflow to run
 GCS_OUTPUT_BUCKET="$4"  # Destination for artifacts
 
 function generate_new_version() {
-  local VERSION=`/versiongenerator --token-file-path=${GITHUB_ACCESS_TOKEN} --org=${REPO_OWNER} --repo=${REPO_NAME}`
+  local VERSION_OUT=`/versiongenerator --token-file-path=${GITHUB_ACCESS_TOKEN} --org=${REPO_OWNER} --repo=${REPO_NAME}`
 
   if [[ $? -ne 0 ]]; then
```
Comment (hopkiw) [inline container_images/daisy-builder/main.sh]: $? now will refer to the success of the VERSION_OUT= assignment, and will probably always be true.

File: container_images/daisy-builder/main.sh [hunk 7]
```diff
@@ -21,13 +21,13 @@ WORKFLOW_FILE="$3"      # Workflow to run
 GCS_OUTPUT_BUCKET="$4"  # Destination for artifacts
 
 function generate_new_version() {
-  local VERSION=`/versiongenerator --token-file-path=${GITHUB_ACCESS_TOKEN} --org=${REPO_OWNER} --repo=${REPO_NAME}`
+  local VERSION_OUT=`/versiongenerator --token-file-path=${GITHUB_ACCESS_TOKEN} --org=${REPO_OWNER} --repo=${REPO_NAME}`
```
Comment (hopkiw) [inline container_images/daisy-builder/main.sh]: does versiongenerator print errors to stdout or stderr?

Comment (iamsubratp) [reply container_images/daisy-builder/main.sh]: yes, it does write errors for stdout

File: container_images/daisy-builder/main.sh (hunk 6 above)
Comment (iamsubratp) [reply container_images/daisy-builder/main.sh]: changed it to catch the error code correctly

Comment (hopkiw) [reply container_images/daisy-builder/main.sh]: Sorry, I wasn't clear. $? will be set to the value of the assignment, which is typically the value of the right operand. But this statement is actually not an assignment, but a `local`. This form of variable initialization always returns 0.

Comment (hopkiw) [reply container_images/daisy-builder/main.sh]: Consider changing to:
```
local VERSION_OUT
if ! VERSION_OUT=`/versiongenerator --token-file-path=${GITHUB_ACCESS_TOKEN} --org=${REPO_OWNER} --repo=${REPO_NAME} 2>&1`; then
    echo "could not generate version: ${VERSION_OUT}"
    return 1
fi
echo $VERSION_OUT
```

Comment (iamsubratp) [reply container_images/daisy-builder/main.sh]: changed

**Pull request merged**